UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,731 @@
+          Serial->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_SERIAL_PORT_INFO);
+          Serial->RegisterStride  = 1;
+          Serial->UseMmio         = 1;
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "current-speed", &TempLen);
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            DEBUG ((DEBUG_INFO, "\n         Property(%08X)  %a", Property, TempStr));
+            DEBUG ((DEBUG_INFO, "  %X", Fdt32ToCpu (*Data32)));
+            Serial->BaudRate = Fdt32ToCpu (*Data32);
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "compatible", &TempLen);
+          TempStr     = (CHAR8 *)(PropertyPtr->Data);
+          if (AsciiStrnCmp (TempStr, "isa", AsciiStrLen ("isa")) == 0) {
+            DEBUG ((DEBUG_INFO, " find serial compatible isa \n"));
+            Serial->UseMmio = 0;
+            PropertyPtr     = FdtGetProperty (Fdt, SubNode, "reg", &TempLen);
+            ASSERT (TempLen > 0);
+            if (TempLen > 0) {
+              Data32               = (UINT32 *)(PropertyPtr->Data);
+              Serial->RegisterBase = Fdt32ToCpu (*(Data32 + 1));
+            }
+          } else {
+            DEBUG ((DEBUG_INFO, " NOT  serial compatible isa \n"));
+            PropertyPtr = FdtGetProperty (Fdt, SubNode, "reg", &TempLen);
+            ASSERT (TempLen > 0);
+            if (TempLen > 0) {
+              Data32               = (UINT32 *)(PropertyPtr->Data);
+              Serial->RegisterBase = Fdt32ToCpu (*Data32);
+            }
+          }
+        }
+      }
+
+      for (Property = FdtFirstPropertyOffset (Fdt, Node); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {
+        PropertyPtr = FdtGetPropertyByOffset (Fdt, Property, &TempLen);
+        TempStr     = FdtGetString (Fdt, Fdt32ToCpu (PropertyPtr->NameOffset), NULL);
+
+        if (AsciiStrCmp (TempStr, "ranges") == 0) {
+          DEBUG ((DEBUG_INFO, "  Found ranges Property TempLen (%08X)\n", TempLen));
+
+          PciRootBridgeInfo->RootBridge[index].AllocationAttributes = EFI_PCI_HOST_BRIDGE_COMBINE_MEM_PMEM | EFI_PCI_HOST_BRIDGE_MEM64_DECODE;
+          PciRootBridgeInfo->RootBridge[index].Supports             = ROOT_BRIDGE_SUPPORTS_DEFAULT;
+          PciRootBridgeInfo->RootBridge[index].PMemAbove4G.Base     = PcdGet64 (PcdPciReservedPMemAbove4GBBase);
+          PciRootBridgeInfo->RootBridge[index].PMemAbove4G.Limit    = PcdGet64 (PcdPciReservedPMemAbove4GBLimit);
+          PciRootBridgeInfo->RootBridge[index].PMem.Base            = PcdGet32 (PcdPciReservedPMemBase);
+          PciRootBridgeInfo->RootBridge[index].PMem.Limit           = PcdGet32 (PcdPciReservedPMemLimit);
+          PciRootBridgeInfo->RootBridge[index].UID                  = index;
+          PciRootBridgeInfo->RootBridge[index].HID                  = EISA_PNP_ID (0x0A03);
+
+          Data32 = (UINT32 *)(PropertyPtr->Data);
+          for (Base = 0; Base < 16; Base = Base + 7) {

```

Comment:
The code does not explain what are these magic numbers 16 and 7 etc. I had shared a working code for RB and this new implementation does not work for me. Trying to understand the implementation..

Reply:
We will update this code. just some clarification.
Currently each address descriptor inside ranges property having 7 elements: addr_type, child_addr_hi, child_addr_low, parent_addr_hi, parent_addr_low, len_hi, len_low, that's why if we wanting to get next descriptor addr_type, we will add 7. Maybe use macro (DWORDS_TO_NEXT_ADDR_TYPE?) for now and figure out optimization in the future.
However, the "Base < 16" should not be hardcoded, I think it should be something similar as "Base <= (TempLen - DWORDS_TO_NEXT_ADDR_TYPE)", we will check this and update.

Reply:
resolved